packages/dds/tree/src/test/forestTestSuite.ts

```diff
@@ -0,0 +1,724 @@
+                                {
+                                    type: Delta.MarkType.MoveInAndModify,
+                                    moveId,
+                                    fields: new Map([
+                                        [
+                                            newField,
+                                            [
+                                                {
+                                                    type: Delta.MarkType.Insert,
+                                                    content: [
+                                                        { type: jsonNumber.name, value: 2 },
+                                                    ].map(singleTextCursor),
+                                                },
+                                            ],
+                                        ],
+                                    ]),
+                                },
+                            ],
+                        ],
+                    ]),
+                };
+                const delta: Delta.Root = new Map([[rootFieldKeySymbol, [mark]]]);
+                // TODO: make type-safe
+                forest.applyDelta(delta);
+
+                const reader = forest.allocateCursor();
+                moveToDetachedField(forest, reader);
+                assert(reader.firstNode());
+                reader.enterField(xField);
+                assert.equal(reader.getFieldLength(), 0);
+                reader.exitField();
+                reader.enterField(yField);
+                assert.equal(reader.getFieldLength(), 2);
+                assert(reader.firstNode());
+                assert.equal(reader.value, 0);
+                reader.enterField(newField);
+                assert.equal(reader.getFieldLength(), 1);
+                assert(reader.firstNode());
+                assert.equal(reader.value, 2);
+            });
+        });
+
+        describe("top level invalidation", () => {
+            it("data editing", () => {
+                const forest = factory(new InMemoryStoredSchemaRepository(defaultSchemaPolicy));
+                const dependent = new MockDependent("dependent");
+                recordDependency(dependent, forest);
+
+                const content: JsonableTree[] = [{ type: jsonNumber.name, value: 1 }];
+                const insert: Delta.Insert = {
+                    type: Delta.MarkType.Insert,
+                    content: content.map(singleTextCursor),
+                };
+                // TODO: make type-safe
+                const delta: Delta.Root = new Map([[rootFieldKeySymbol, [insert]]]);
+
+                assert.deepEqual(dependent.tokens, []);
+                forest.applyDelta(delta);
+                assert.deepEqual(dependent.tokens.length, 1);
+
+                forest.applyDelta(delta);
+                assert.deepEqual(dependent.tokens.length, 2);
+
+                // TODO: maybe test some other deltas.
```

Comment:
I don't think adding tests for other deltas would add much value.
Is there a way to remove the dependency so that the dependent stops getting inval messages? If so it would be good to test that.